File: services/core/src/cortex_core/synthesis.py
"""At-most-one provider attempt per personal key; ambiguous attempts never replay."""

import re
from uuid import uuid4

from .auth import CoreError
from .companion_responses import CompanionResponseService
from .contracts import CompanionResponseInput, SynthesisUsage
from .model_attempts import ModelAttemptService
from .service import digest, encoded, one, run
from .workspace import WorkspaceService


class SynthesisService:
    def __init__(self, knowledge, factory=None, daily_limit=100):
        self.k, self.db, self.factory = knowledge, knowledge.db, factory
        self.quota = ModelAttemptService(knowledge, daily_limit)
        self.responses = CompanionResponseService(knowledge)

    def _lock(self, conn, p, domain, *, write=False):
        run(
            conn,
            "SELECT id FROM cf_domains WHERE tenant_id=:tenant AND id=:domain "
            + ("FOR UPDATE" if write else "FOR KEY SHARE"),
            **self.k.keys(p, domain),
        )
        if not one(
            conn,
            "SELECT role FROM cf_memberships WHERE tenant_id=:tenant AND domain_id=:domain AND subject=:subject",
            **self.k.keys(p, domain),
            subject=p.subject,
        ):
            raise CoreError("NOT_FOUND", "Domain not found", 404)

    def _attempt(self, conn, p, domain, ident):
        row = one(
            conn,
            "SELECT * FROM cf_synthesis_attempts WHERE tenant_id=:tenant AND domain_id=:domain AND id=:id AND subject=:subject",
            **self.k.keys(p, domain),
            id=ident,
            subject=p.subject,
        )
        if not row:
            raise CoreError("NOT_FOUND", "Synthesis not found", 404)
        self.k._episode(conn, p, domain, row["episode_id"])
        return row

    def _view(self, conn, p, domain, row):
        outcome = one(
            conn,
            "SELECT * FROM cf_synthesis_outcomes WHERE tenant_id=:tenant AND domain_id=:domain AND attempt_id=:id",
            **self.k.keys(p, domain),
            id=row["id"],
        )
        return {
            **{
                key: row[key]
                for key in (
                    "id",
                    "episode_id",
                    "provider",
                    "requested_model",
                    "prompt_version",
                    "budget_reserved",
                    "idempotency_key",
                    "created_at",
                )
            },
            "status": outcome["status"] if outcome else "unresolved",
            "response_id": outcome["response_id"] if outcome else None,
            "error_code": outcome["error_code"] if outcome else None,
            "usage": outcome["usage"] if outcome else None,
            "finished_at": outcome["created_at"] if outcome else None,
        }

    def detail(self, p, domain, ident):
        with self.db.transaction(p, domain, isolation="READ COMMITTED") as conn:
            self._lock(conn, p, domain)
            return self._view(conn, p, domain, self._attempt(conn, p, domain, ident))

    def listing(self, p, domain, limit, after, episode_id=None, key=None):
        with self.db.transaction(p, domain, isolation="READ COMMITTED") as conn:
            self._lock(conn, p, domain)
            if episode_id:
                self.k._episode(conn, p, domain, episode_id)
            return WorkspaceService(self.k)._page(
                conn,
                p,
                domain,
                "cf_synthesis_attempts",
                limit,
                after,
                lambda row: self.k._episode(conn, p, domain, row["episode_id"]),
                lambda row: self._view(conn, p, domain, row),
                "AND subject=:subject AND (:episode='' OR episode_id=:episode) AND (:key='' OR idempotency_key=:key)",
                {"subject": p.subject, "episode": episode_id or "", "key": key or ""},
            )

    @staticmethod
    def _identity(p, authenticate):
        current = authenticate()
        if current != p:
            raise CoreError("NOT_AUTHORIZED", "Identity changed during synthesis", 403)

    def _reserve(self, p, domain, episode_id, data):
        fingerprint = digest({"episode_id": episode_id, **data.model_dump(mode="json")})
        with self.db.transaction(p, domain, isolation="READ COMMITTED") as conn:
            self._lock(conn, p, domain, write=True)
            episode = self.k._episode(conn, p, domain, episode_id)
            old = one(
                conn,
                "SELECT * FROM cf_synthesis_attempts WHERE tenant_id=:tenant AND domain_id=:domain AND subject=:subject AND idempotency_key=:key",
                **self.k.keys(p, domain),
                subject=p.subject,
                key=data.idempotency_key,
            )
            if old:
                if old["request_hash"] != fingerprint:
                    raise CoreError("IDEMPOTENCY_CONFLICT", "Synthesis key reused")
                return old["id"], None, False
            if self.factory is None:
                raise CoreError("SYNTHESIS_DISABLED", "Backend synthesis is not configured", 503)
            model = self.factory()
            paid = bool(episode["result"]["citations"])
            prepared = (
                model.prepare(episode["question"], episode["result"])
                if paid
                else {"abstention": episode_id}
            )
            if paid and self.quota._usage(conn, p, domain)["remaining_attempts"] == 0:
                raise CoreError(
                    "MODEL_DAILY_LIMIT", "The domain model attempt allowance is exhausted", 429
                )
            ident = str(uuid4())
            run(
                conn,
                """INSERT INTO cf_synthesis_attempts(tenant_id,domain_id,id,subject,episode_id,provider,requested_model,prompt_version,budget_reserved,input_sha256,idempotency_key,request_hash)
                VALUES(:tenant,:domain,:id,:subject,:episode,:provider,:model,:prompt,:paid,:input,:key,:hash)""",
                **self.k.keys(p, domain),
                id=ident,
                subject=p.subject,
                episode=episode_id,
                provider="openrouter" if paid else "none",
                model=model.model if paid else None,
                prompt=model.PROMPT_VERSION if paid else "deterministic-abstention-v1",
                paid=paid,
                input=digest(prepared),
                key=data.idempotency_key,
                hash=fingerprint,
            )
            # Also invalidate repeatable-read extraction quota snapshots.
            run(
                conn,
                "UPDATE cf_domains SET accepted_version=accepted_version WHERE tenant_id=:tenant AND id=:domain",
                **self.k.keys(p, domain),
            )
            return ident, model if paid else None, True

    def _failed(self, p, domain, ident, exc, usage):
        code = (
            exc.code
            if isinstance(exc, CoreError) and re.fullmatch(r"[A-Z_]{1,100}", exc.code)
            else "SYNTHESIS_FAILED"
        )
        # An in-flight attempt may outlive membership revocation. This internal
        # append records only a safe outcome for its original authenticated author.
        with self.db.engine.begin() as conn:
            run(conn, "SELECT set_config('cortex.tenant',:tenant,true)", tenant=p.tenant_id)
            run(
                conn,
                """INSERT INTO cf_synthesis_outcomes(tenant_id,domain_id,attempt_id,status,error_code,usage)
                SELECT tenant_id,domain_id,id,'failed',:code,CAST(:usage AS jsonb) FROM cf_synthesis_attempts
                WHERE tenant_id=:tenant AND domain_id=:domain AND id=:id AND subject=:subject
                ON CONFLICT(tenant_id,domain_id,attempt_id) DO NOTHING""",
                **self.k.keys(p, domain),
                id=ident,
                subject=p.subject,
                code=code,
                usage=encoded(usage),
            )

    def execute(self, p, domain, episode_id, data, authenticate):
        self._identity(p, authenticate)
        ident, model, fresh = self._reserve(p, domain, episode_id, data)
        if not fresh:
            self._identity(p, authenticate)
            return self.detail(p, domain, ident)
        usage = None
        try:
            self._identity(p, authenticate)
            with self.db.transaction(p, domain, isolation="READ COMMITTED") as conn:
                self._lock(conn, p, domain)
                episode = self.k._episode(conn, p, domain, episode_id)
            if model:
                draft = model.synthesize(episode["question"], episode["result"])
                usage = SynthesisUsage.model_validate(draft["usage"]).model_dump(mode="json")
            else:
                draft = {
                    "answer_text": "Aucune preuve exploitable dans cet épisode ; je ne peux pas fournir une réponse sourcée.",
                    "answer_kind": "abstention",
                    "citations": [],
                    "model": None,
                }
            self._identity(p, authenticate)
            with self.db.transaction(p, domain, isolation="READ COMMITTED") as conn:
                self._lock(conn, p, domain, write=True)
                self._attempt(conn, p, domain, ident)
                receipt = self.responses._create(
                    conn,
                    p,
                    domain,
                    episode_id,
                    CompanionResponseInput(
                        **{
                            key: draft[key]
                            for key in ("answer_text", "answer_kind", "citations", "model")
                        },
                        companion="cortex-backend-synthesis-v1",
                        idempotency_key="synthesis:" + ident,
                    ),
                )
                run(
                    conn,
                    """INSERT INTO cf_synthesis_outcomes(tenant_id,domain_id,attempt_id,status,response_id,usage)
                    VALUES(:tenant,:domain,:id,'succeeded',:response,CAST(:usage AS jsonb))""",
                    **self.k.keys(p, domain),
                    id=ident,
                    response=receipt["id"],
                    usage=encoded(usage),
                )
        except Exception as exc:
            try:
                if usage is None and model and getattr(model, "last_usage", None):
                    usage = SynthesisUsage.model_validate(model.last_usage).model_dump(mode="json")
            except ValueError:
                usage = None
            try:
                self._failed(p, domain, ident, exc, usage)
            except Exception:
                raise CoreError(
                    "SYNTHESIS_STORAGE_UNCERTAIN",
                    "Inspect the existing attempt; do not automatically retry with a new key",
                    503,
                ) from None
            if isinstance(exc, CoreError) and exc.status in (401, 403, 404):
                raise
        self._identity(p, authenticate)
        return self.detail(p, domain, ident)

File: services/rust-core/src/synthesis.rs
//! Personal durable synthesis attempts. No retry of an existing or ambiguous reservation.
use crate::{
    auth::Principal, companions, error::CoreError, model_provider, retrieval, server::StateData,
};
use axum::{
    Json, Router,
    extract::{Path, State},
    routing::post,
};
use http::{HeaderMap, StatusCode};
use serde::Deserialize;
use serde_json::{Value, json};
use sqlx::{Postgres, Row, Transaction};
use uuid::Uuid;
fn invalid() -> CoreError {
    model_provider::error("VALIDATION_FAILED", StatusCode::UNPROCESSABLE_ENTITY)
}
#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Input {
    processing_destination: String,
    idempotency_key: String,
}
pub fn routes() -> Router<StateData> {
    Router::new().route(
        "/v1/domains/{domain}/episodes/{episode_id}/syntheses",
        post(execute),
    )
}
async fn episode(
    tx: &mut Transaction<'_, Postgres>,
    p: &Principal,
    d: &str,
    id: &str,
) -> Result<sqlx::postgres::PgRow, CoreError> {
    let sources = retrieval::accessible(tx, p, d).await?;
    retrieval::episode_row(tx, p, d, id, &sources).await
}
async fn receipt(
    tx: &mut Transaction<'_, Postgres>,
    p: &Principal,
    d: &str,
    id: &str,
) -> Result<Value, CoreError> {
    let r=sqlx::query("SELECT a.*,to_jsonb(a.created_at) AS created,COALESCE(o.status,'unresolved') AS outcome_status,to_jsonb(o.created_at) AS finished,o.error_code,o.response_id,o.usage FROM cf_synthesis_attempts a LEFT JOIN cf_synthesis_outcomes o ON o.tenant_id=a.tenant_id AND o.domain_id=a.domain_id AND o.attempt_id=a.id WHERE a.tenant_id=$1 AND a.domain_id=$2 AND a.id=$3 AND a.subject=$4").bind(&p.tenant).bind(d).bind(id).bind(&p.subject).fetch_optional(&mut **tx).await.map_err(CoreError::sql)?.ok_or_else(CoreError::not_found)?;
    episode(tx, p, d, r.get("episode_id")).await?;
    p.check_fresh()?;
    Ok(
        json!({"id":id,"episode_id":r.get::<String,_>("episode_id"),"provider":r.get::<String,_>("provider"),"requested_model":r.get::<Option<String>,_>("requested_model"),"prompt_version":r.get::<String,_>("prompt_version"),"budget_reserved":r.get::<bool,_>("budget_reserved"),"idempotency_key":r.get::<String,_>("idempotency_key"),"created_at":r.get::<Value,_>("created"),"status":r.get::<String,_>("outcome_status"),"response_id":r.get::<Option<String>,_>("response_id"),"error_code":r.get::<Option<String>,_>("error_code"),"usage":r.get::<Option<Value>,_>("usage"),"finished_at":r.get::<Option<Value>,_>("finished")}),
    )
}
pub(crate) async fn quota(
    tx: &mut Transaction<'_, Postgres>,
    p: &Principal,
    d: &str,
    limit: i64,
) -> Result<(), CoreError> {
    let count:i64=sqlx::query_scalar("SELECT count(*) FROM (SELECT created_at FROM cf_model_attempts WHERE tenant_id=$1 AND domain_id=$2 UNION ALL SELECT created_at FROM cf_synthesis_attempts WHERE tenant_id=$1 AND domain_id=$2 AND budget_reserved) q WHERE created_at>=date_trunc('day',now() AT TIME ZONE 'UTC') AT TIME ZONE 'UTC'").bind(&p.tenant).bind(d).fetch_one(&mut **tx).await.map_err(CoreError::sql)?;
    if count >= limit {
        return Err(model_provider::error(
            "MODEL_DAILY_LIMIT",
            StatusCode::TOO_MANY_REQUESTS,
        ));
    }
    Ok(())
}
async fn failed(
    s: &StateData,
    p: &Principal,
    d: &str,
    id: &str,
    code: &str,
    usage: &Option<Value>,
) -> Result<(), CoreError> {
    let mut tx = s.db.pool.begin().await.map_err(CoreError::sql)?;
    sqlx::query("SELECT set_config('cortex.tenant',$1,true)")
        .bind(&p.tenant)
        .execute(&mut *tx)
        .await
        .map_err(CoreError::sql)?;
    sqlx::query("INSERT INTO cf_synthesis_outcomes(tenant_id,domain_id,attempt_id,status,error_code,usage) SELECT tenant_id,domain_id,id,'failed',$5,$6 FROM cf_synthesis_attempts WHERE tenant_id=$1 AND domain_id=$2 AND id=$3 AND subject=$4 ON CONFLICT(tenant_id,domain_id,attempt_id) DO NOTHING").bind(&p.tenant).bind(d).bind(id).bind(&p.subject).bind(code).bind(usage).execute(&mut *tx).await.map_err(CoreError::sql)?;
    tx.commit().await.map_err(CoreError::sql)
}
async fn unconfirmed_execute(
    State(s): State<StateData>,
    Path((d, e)): Path<(String, String)>,
    h: HeaderMap,
    input: Result<Json<Input>, axum::extract::rejection::JsonRejection>,
) -> Result<Json<Value>, CoreError> {
    let p = s.auth.authenticate(&h)?;
    let d = Uuid::parse_str(&d)
        .map_err(|_| CoreError::invalid_uuid())?
        .to_string();
    let e = Uuid::parse_str(&e)
        .map_err(|_| CoreError::invalid_uuid())?
        .to_string();
    let Json(i) = input.map_err(|_| invalid())?;
    if i.processing_destination != "openrouter"
        || !(8..=128).contains(&i.idempotency_key.chars().count())
    {
        return Err(invalid());
    }
    let fingerprint=crate::canonical::digest(&json!({"episode_id":e,"processing_destination":i.processing_destination,"idempotency_key":i.idempotency_key})).map_err(|_|invalid())?;
    let mut tx = retrieval::transaction(&s, &p, &d).await?;
    let ep = episode(&mut tx, &p, &d, &e).await?;
    let old=sqlx::query("SELECT id,request_hash FROM cf_synthesis_attempts WHERE tenant_id=$1 AND domain_id=$2 AND subject=$3 AND idempotency_key=$4").bind(&p.tenant).bind(&d).bind(&p.subject).bind(&i.idempotency_key).fetch_optional(&mut *tx).await.map_err(CoreError::sql)?;
    if let Some(r) = old {
        if r.get::<String, _>("request_hash") != fingerprint {
            return Err(model_provider::error(
                "IDEMPOTENCY_CONFLICT",
                StatusCode::CONFLICT,
            ));
        }
        let result = receipt(&mut tx, &p, &d, r.get("id")).await?;
        tx.commit().await.map_err(CoreError::sql)?;
        return Ok(Json(result));
    }
    let model = s.synthesis.as_ref().ok_or_else(|| {
        model_provider::error("SYNTHESIS_DISABLED", StatusCode::SERVICE_UNAVAILABLE)
    })?;
    let result: Value = ep.get("result");
    let paid = !result["citations"]
        .as_array()
        .ok_or_else(CoreError::database)?
        .is_empty();
    let prepared = if paid {
        model.prepare(ep.get("question"), &result)?
    } else {
        json!({"abstention":e})
    };
    if paid {
        quota(&mut tx, &p, &d, s.model_daily_limit).await?
    }
    let id = Uuid::new_v4().to_string();
    let input_hash = crate::canonical::digest(&prepared).map_err(|_| invalid())?;
    sqlx::query("INSERT INTO cf_synthesis_attempts(tenant_id,domain_id,id,subject,episode_id,provider,requested_model,prompt_version,budget_reserved,input_sha256,idempotency_key,request_hash) VALUES($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12)").bind(&p.tenant).bind(&d).bind(&id).bind(&p.subject).bind(&e).bind(if paid{"openrouter"}else{"none"}).bind(if paid{Some(&model.model)}else{None}).bind(if paid{"cited-synthesis-v2"}else{"deterministic-abstention-v1"}).bind(paid).bind(input_hash).bind(&i.idempotency_key).bind(&fingerprint).execute(&mut *tx).await.map_err(CoreError::sql)?;
    sqlx::query(
        "UPDATE cf_domains SET accepted_version=accepted_version WHERE tenant_id=$1 AND id=$2",
    )
    .bind(&p.tenant)
    .bind(&d)
    .execute(&mut *tx)
    .await
    .map_err(CoreError::sql)?;
    p.check_fresh()?;
    tx.commit().await.map_err(CoreError::sql)?;
    let mut usage = None;
    let outcome:Result<(),CoreError>=async {
 p.check_fresh()?;let mut tx=retrieval::transaction(&s,&p,&d).await?;episode(&mut tx,&p,&d,&e).await?;tx.commit().await.map_err(CoreError::sql)?;
 let mut draft=if paid{let response=model.request(&prepared).await?;let (model_name,u)=model_provider::usage(&response)?;usage=Some(u);model_provider::draft(&response,&result,&model_name)?}else{json!({"answer_text":"Aucune preuve exploitable dans cet épisode ; je ne peux pas fournir une réponse sourcée.","answer_kind":"abstention","citations":[],"model":null})};
 p.check_fresh()?;let mut tx=retrieval::transaction(&s,&p,&d).await?;episode(&mut tx,&p,&d,&e).await?;draft["companion"]=json!("cortex-backend-synthesis-v1");draft["idempotency_key"]=json!(format!("synthesis:{id}"));let input:companions::CompanionInput=serde_json::from_value(draft).map_err(|_|model_provider::error("UNSUPPORTED_SYNTHESIS",StatusCode::UNPROCESSABLE_ENTITY))?;let response=companions::create_in_transaction(&mut tx,&p,&d,&e,input).await?;
 sqlx::query("INSERT INTO cf_synthesis_outcomes(tenant_id,domain_id,attempt_id,status,response_id,usage) VALUES($1,$2,$3,'succeeded',$4,$5)").bind(&p.tenant).bind(&d).bind(&id).bind(response["id"].as_str().ok_or_else(CoreError::database)?).bind(&usage).execute(&mut *tx).await.map_err(CoreError::sql)?;p.check_fresh()?;tx.commit().await.map_err(CoreError::sql)?;Ok(())}.await;
    if let Err(error) = outcome {
        if failed(&s, &p, &d, &id, error.code, &usage).await.is_err() {
            return Err(model_provider::error(
                "SYNTHESIS_STORAGE_UNCERTAIN",
                StatusCode::SERVICE_UNAVAILABLE,
            ));
        }
        if matches!(error.status.as_u16(), 401 | 403 | 404) {
            return Err(error);
        }
    }
    p.check_fresh()?;
    let mut tx = retrieval::transaction(&s, &p, &d).await?;
    let result = receipt(&mut tx, &p, &d, &id).await?;
    tx.commit().await.map_err(CoreError::sql)?;
    Ok(Json(result))
}

async fn execute(
    State(s): State<StateData>,
    Path((d, e)): Path<(String, String)>,
    h: HeaderMap,
    axum::extract::RawQuery(q): axum::extract::RawQuery,
    proof: Option<axum::Extension<crate::confirmation::ConfirmedAction>>,
    body: axum::body::Bytes,
) -> axum::response::Response {
    use axum::response::IntoResponse;
    let action = "syntheses.create";
    let raw = serde_json::from_slice::<Value>(&body);
    let args =
        json!({"path":{"domain":d,"episode_id":e},"body":raw.as_ref().unwrap_or(&Value::Null)});
    let outcome: Result<Json<Value>, CoreError> = async {
        let p = s.auth.authenticate(&h)?;
        let domain = Uuid::parse_str(&d)
            .map_err(|_| CoreError::invalid_uuid())?
            .to_string();
        Uuid::parse_str(&e).map_err(|_| CoreError::invalid_uuid())?;
        let i: Input =
            serde_json::from_value(raw.map_err(|_| invalid())?).map_err(|_| invalid())?;
        if q.is_some_and(|q| !q.is_empty())
            || i.processing_destination != "openrouter"
            || !(8..=128).contains(&i.idempotency_key.chars().count())
        {
            return Err(invalid());
        }
        let tx = crate::confirmation::authorize(&s.db, &p, &domain, action).await?;
        tx.commit().await.map_err(CoreError::sql)?;
        if !proof
            .as_ref()
            .is_some_and(|v| v.subject == p.subject && v.tenant == p.tenant && v.action == action)
        {
            if h.get_all("x-cortex-confirmation").iter().count() > 1 {
                return Err(invalid());
            };
            s.confirmation
                .as_ref()
                .ok_or_else(crate::confirmation::required)?
                .consume(
                    &p,
                    &domain,
                    action,
                    &args,
                    h.get("x-cortex-confirmation").and_then(|v| v.to_str().ok()),
                )
                .await?;
        }
        unconfirmed_execute(State(s), Path((d, e)), h, Ok(Json(i))).await
    }
    .await;
    match outcome{Ok(v)=>v.into_response(),Err(e)if e.code=="CONFIRMATION_REQUIRED"=>(e.status,Json(json!({"error":e.code,"message":e.message,"confirmation_request":{"action":action,"command_hash":crate::confirmation::command_hash(action,&args).unwrap_or_default(),"transport_header":"X-Cortex-Confirmation","max_lifetime_seconds":300}}))).into_response(),Err(e)=>e.into_response()}
}
